src/not-so-smart/decoder.ml

```diff
@@ -124,34 +124,31 @@ let at_least_one_line decoder =
   done;
   !pos < decoder.max && !chr = '\n' && !has_cr
 
-let digit = function
-  | 'a' .. 'f' as chr -> 10 + Char.code chr - Char.code 'a'
-  | 'A' .. 'F' as chr -> 10 + Char.code chr - Char.code 'A'
-  | '0' .. '9' as chr -> Char.code chr - Char.code '0'
-  | _ -> invalid_arg "invalid digit"
-
-let to_int ~base ~off ~len buf =
-  let code = ref 0 in
-  for i = 0 to len - 1 do
-    let v = digit (Bytes.get buf (off + i)) in
-    assert (v < base);
-    code := (base * !code) + v
-  done;
-  !code
+(** reads off 4 bytes from [decoder.buffer] starting at [decoder.pos] and interprets read
+    bytes as hex and converts to int.
+    Why unsafe:
+    @raise Invalid_argument if there are no 4 bytes to read, i.e.,
+           [decoder.max - decoder.pos < 4]  *)
+let pkt_len_unsafe (decoder : decoder) =
+  let hex = Bytes.of_string "0x0000" in
+  Bytes.blit decoder.buffer decoder.pos hex 2 4;
+  Bytes.to_string hex |> int_of_string
```

Comment:
```ocaml
int_of_string (Bytes.to_string hex)
```

is better and you can safely use `Bytes.unsafe_to_string` in that case to avoid a double-allocation.

Reply:
just to be clear:
1. you don't want me to use `|>`? 
2. from your message it seems like `|>` prevents us from using `unsafe_to_string`. If that's true, why is it so?